diff --git a/Makefile b/Makefile
index afe325c16..3f2e19415 100644
--- a/Makefile
+++ b/Makefile
@@ -709,8 +709,12 @@ drivers-y	:= drivers/ sound/
 ifdef CONFIG_AMLOGIC_IN_KERNEL_MODULES
 drivers-y	+= $(COMMON_DRIVERS_DIR)/drivers/ $(COMMON_DRIVERS_DIR)/sound/
 drivers-y       += $(COMMON_DRIVERS_DIR)/samples/
+ifdef CONFIG_CC_IS_CLANG
+KBUILD_CFLAGS += -Wno-misleading-indentation -Wno-error=address -Wno-enum-conversion -Warray-bounds
+else
 KBUILD_CFLAGS += -Wno-misleading-indentation -Wno-error=address -Wno-enum-int-mismatch -Wno-frame-larger-than -Wno-stringop-overread
 endif
+endif
 drivers-$(CONFIG_SAMPLES) += samples/
 drivers-$(CONFIG_NET) += net/
 drivers-y	+= virt/
diff --git a/common_drivers/drivers/amlisp/amlcam/Makefile b/common_drivers/drivers/amlisp/amlcam/Makefile
index 0c5b39e50..573fca59f 100644
--- a/common_drivers/drivers/amlisp/amlcam/Makefile
+++ b/common_drivers/drivers/amlisp/amlcam/Makefile
@@ -5,6 +5,9 @@ M_PATH := $(shell dirname $(lastword $(MAKEFILE_LIST)))
 
 ifeq ($(DRV_SRC_OBJ),)
 	DRV_SRC := $(wildcard $(M_PATH)/*.c $(M_PATH)/cam_common/*.c $(M_PATH)/cam_isp/*.c $(M_PATH)/cam_isp/*/*.c $(M_PATH)/cam_adapter/*.c $(M_PATH)/cam_adapter/*/*.c $(M_PATH)/cam_csiphy/*.c $(M_PATH)/cam_csiphy/*/*.c $(M_PATH)/cam_pattern/*.c)
+ifdef CONFIG_CC_IS_CLANG
+	DRV_SRC := $(filter-out %.mod.c,$(DRV_SRC)) 
+endif
 	DRV_SRC := $(subst $(M_PATH)/,,$(DRV_SRC))
 	export DRV_SRC_OBJ := $(DRV_SRC:.c=.o)
 endif
diff --git a/common_drivers/drivers/amlisp/amllens/dw9714/Makefile b/common_drivers/drivers/amlisp/amllens/dw9714/Makefile
index 202f29ee5..54b17349c 100644
--- a/common_drivers/drivers/amlisp/amllens/dw9714/Makefile
+++ b/common_drivers/drivers/amlisp/amllens/dw9714/Makefile
@@ -4,8 +4,12 @@ M_PATH := $(shell dirname $(lastword $(MAKEFILE_LIST)))
 
 DRV_SRC := $(wildcard $(M_PATH)/*.c)
 
+ifdef CONFIG_CC_IS_CLANG
+DRV_SRC := $(filter-out %.mod.c,$(DRV_SRC)) 
+endif
+
 M_DRV_SRC := $(DRV_SRC:$(M_PATH)/%=%)
 
 DRV_SRC_OBJ := $(M_DRV_SRC:.c=.o)
 
-dw9714-objs := $(DRV_SRC_OBJ)
\ No newline at end of file
+dw9714-objs := $(DRV_SRC_OBJ)
diff --git a/common_drivers/drivers/amlisp/amllens/dw9800w/Makefile b/common_drivers/drivers/amlisp/amllens/dw9800w/Makefile
index 4664f4c36..fe361f5f2 100644
--- a/common_drivers/drivers/amlisp/amllens/dw9800w/Makefile
+++ b/common_drivers/drivers/amlisp/amllens/dw9800w/Makefile
@@ -5,6 +5,10 @@ M_PATH := $(shell dirname $(lastword $(MAKEFILE_LIST)))
 
 DRV_SRC := $(wildcard $(M_PATH)/*.c)
 
+ifdef CONFIG_CC_IS_CLANG
+DRV_SRC := $(filter-out %.mod.c,$(DRV_SRC)) 
+endif
+
 M_DRV_SRC := $(DRV_SRC:$(M_PATH)/%=%)
 
 DRV_SRC_OBJ := $(M_DRV_SRC:.c=.o)
diff --git a/common_drivers/drivers/amlisp/amlsens/Makefile b/common_drivers/drivers/amlisp/amlsens/Makefile
index fb6f4bdff..17d9b77e7 100644
--- a/common_drivers/drivers/amlisp/amlsens/Makefile
+++ b/common_drivers/drivers/amlisp/amlsens/Makefile
@@ -6,6 +6,9 @@ M_PATH := $(shell dirname $(lastword $(MAKEFILE_LIST)))
 ifeq ($(DRV_SRC_OBJ),)
 	DRV_SRC := $(wildcard $(M_PATH)/*.c $(M_PATH)/imx290/*.c $(M_PATH)/imx335/*.c $(M_PATH)/imx378/*.c $(M_PATH)/imx415/*.c $(M_PATH)/ov08a10/*.c $(M_PATH)/ov5640/*.c \
 				$(M_PATH)/ov13b10/*.c $(M_PATH)/imx577/*.c $(M_PATH)/ov16a1q/*c $(M_PATH)/ov13855/*.c $(M_PATH)/lt6911c/*.c)
+ifdef CONFIG_CC_IS_CLANG
+	DRV_SRC := $(filter-out %.mod.c,$(DRV_SRC)) 
+endif
 	DRV_SRC := $(subst $(M_PATH)/,,$(DRV_SRC))
 	export DRV_SRC_OBJ := $(DRV_SRC:.c=.o)
 endif
diff --git a/common_drivers/drivers/media_modules/amvdec_ports/aml_vcodec_dec.c b/common_drivers/drivers/media_modules/amvdec_ports/aml_vcodec_dec.c
index e9814e52c..2638fbdc5 100644
--- a/common_drivers/drivers/media_modules/amvdec_ports/aml_vcodec_dec.c
+++ b/common_drivers/drivers/media_modules/amvdec_ports/aml_vcodec_dec.c
@@ -4621,7 +4621,7 @@ static int vb2ops_vdec_buf_init(struct vb2_buffer *vb)
 	}
 
 	if (!V4L2_TYPE_IS_OUTPUT(vb->type)) {
-		ulong key;
+		ulong key = 0;
 		if (vb->memory == VB2_MEMORY_DMABUF)
 			key = (ulong)vb->planes[0].dbuf;
 		else if (vb->memory == VB2_MEMORY_MMAP)
diff --git a/common_drivers/drivers/media_modules/amvdec_ports/decoder/utils.h b/common_drivers/drivers/media_modules/amvdec_ports/decoder/utils.h
index 45268ee12..fdcc35723 100644
--- a/common_drivers/drivers/media_modules/amvdec_ports/decoder/utils.h
+++ b/common_drivers/drivers/media_modules/amvdec_ports/decoder/utils.h
@@ -26,8 +26,10 @@
 	(((x) > (high)) ? (high) : (((x) < (low)) ? (low) : (x)))
 #define BITAT(x, n) ((x & (1 << n)) == (1 << n))
 
+#ifndef CONFIG_CC_IS_CLANG
 typedef unsigned char uint8_t;
 typedef int int32_t;
 typedef unsigned int uint32_t;
+#endif
 
 #endif //_UTILS_H
diff --git a/common_drivers/drivers/media_modules/frame_provider/decoder/avs2_fb/vavs2_fb.c b/common_drivers/drivers/media_modules/frame_provider/decoder/avs2_fb/vavs2_fb.c
index a5d8d42d6..e7aab93c5 100644
--- a/common_drivers/drivers/media_modules/frame_provider/decoder/avs2_fb/vavs2_fb.c
+++ b/common_drivers/drivers/media_modules/frame_provider/decoder/avs2_fb/vavs2_fb.c
@@ -7474,7 +7474,7 @@ static irqreturn_t vavs2_isr_thread_fn(int irq, void *data)
 				/*avs2_dec->m_bg->index is
 				set to dec->used_buf_num - 1*/
 			if ((dec->front_back_mode == 1) && (paral_alloc_buffer_mode & 1)) {
-				if ((dec->pic_list_wait_alloc_done_flag == BUFFER_INIT)) {
+				if (dec->pic_list_wait_alloc_done_flag == BUFFER_INIT) {
 					dec->dec_result = DEC_RESULT_WAIT_BUFFER;
 					avs2_print(dec, AVS2_DBG_BUFMGR, "alloc buffer\n");
 					ATRACE_COUNTER(dec->trace.decode_time_name, DECODER_ISR_THREAD_HEAD_END);
diff --git a/common_drivers/drivers/media_modules/frame_provider/decoder/avs3/avs3_fb_hw.c b/common_drivers/drivers/media_modules/frame_provider/decoder/avs3/avs3_fb_hw.c
index 6c09654ce..85dbcbe4d 100644
--- a/common_drivers/drivers/media_modules/frame_provider/decoder/avs3/avs3_fb_hw.c
+++ b/common_drivers/drivers/media_modules/frame_provider/decoder/avs3/avs3_fb_hw.c
@@ -2682,7 +2682,7 @@ static void config_dw_fb(struct AVS3Decoder_s *dec, struct avs3_frame_s *pic,
 
 	struct avs3_decoder *avs3_dec = &dec->avs3_dec;
 	int dw_mode = get_double_write_mode(dec);
-	uint32_t data, data32;
+	uint32_t data = 0, data32;
 	if ((dw_mode & 0x10) == 0) {
 		WRITE_BACK_8(avs3_dec, HEVC_SAO_CTRL26, 0);
 
diff --git a/common_drivers/drivers/media_modules/frame_provider/decoder/avs3/vavs3.c b/common_drivers/drivers/media_modules/frame_provider/decoder/avs3/vavs3.c
index 4aa7015c9..7a554e719 100644
--- a/common_drivers/drivers/media_modules/frame_provider/decoder/avs3/vavs3.c
+++ b/common_drivers/drivers/media_modules/frame_provider/decoder/avs3/vavs3.c
@@ -7918,7 +7918,7 @@ static irqreturn_t vavs3_isr_thread_fn(int irq, void *data)
 			int32_t lcu_size_log2 = avs3_dec->lcu_size_log2;
 
 			if ((dec->front_back_mode == 1) && (paral_alloc_buffer_mode & 1)) {
-				if ((dec->pic_list_wait_alloc_done_flag == BUFFER_INIT)) {
+				if (dec->pic_list_wait_alloc_done_flag == BUFFER_INIT) {
 					dec->dec_result = DEC_RESULT_WAIT_BUFFER;
 					avs3_print(dec, AVS3_DBG_BUFMGR, "alloc buffer\n");
 					vdec_schedule_work(&dec->work);
diff --git a/common_drivers/drivers/media_modules/frame_provider/decoder/h265_fb/vh265_fb.c b/common_drivers/drivers/media_modules/frame_provider/decoder/h265_fb/vh265_fb.c
index 813241cea..a7f01a855 100644
--- a/common_drivers/drivers/media_modules/frame_provider/decoder/h265_fb/vh265_fb.c
+++ b/common_drivers/drivers/media_modules/frame_provider/decoder/h265_fb/vh265_fb.c
@@ -3747,7 +3747,7 @@ static void delrefill(struct hevc_state_s *hevc)
 static void get_rpm_param(union param_u *params)
 {
 	int i;
-	unsigned int data32;
+	unsigned int data32 = 0;
 
 	for (i = 0; i < 128; i++) {
 		do {
@@ -4881,7 +4881,7 @@ static void init_decode_head_hw(struct hevc_state_s *hevc)
 {
 
 	struct BuffInfo_s *buf_spec = hevc->work_space_buf;
-	unsigned int data32;
+	unsigned int data32 = 0;
 
 	int losless_comp_header_size =
 		compute_losless_comp_header_size(hevc->pic_w,
diff --git a/common_drivers/drivers/media_modules/frame_provider/decoder/vav1_fb/av1_fb_hw.c b/common_drivers/drivers/media_modules/frame_provider/decoder/vav1_fb/av1_fb_hw.c
index 0db3ce186..786fe2a6a 100644
--- a/common_drivers/drivers/media_modules/frame_provider/decoder/vav1_fb/av1_fb_hw.c
+++ b/common_drivers/drivers/media_modules/frame_provider/decoder/vav1_fb/av1_fb_hw.c
@@ -2576,7 +2576,7 @@ static void config_sao_hw_fb(struct AV1HW_s *hw, param_t* params)
 		READ_WRITE_DATA16(hw, HEVC_SAO_CTRL5, 0, 16, 8);
 
 	} else {
-		uint32_t data;
+		uint32_t data = 0;
 		if (get_cpu_major_id() >= AM_MESON_CPU_MAJOR_ID_T7) {
 			WRITE_BACK_8(hw, HEVC_SAO_CTRL26, 0);
 		}
diff --git a/common_drivers/drivers/media_modules/frame_provider/decoder/vp9_fb/vvp9_fb.c b/common_drivers/drivers/media_modules/frame_provider/decoder/vp9_fb/vvp9_fb.c
index 21e575918..2a30ec73c 100644
--- a/common_drivers/drivers/media_modules/frame_provider/decoder/vp9_fb/vvp9_fb.c
+++ b/common_drivers/drivers/media_modules/frame_provider/decoder/vp9_fb/vvp9_fb.c
@@ -5050,7 +5050,7 @@ static void config_sao_hw_fb(struct VP9Decoder_s *pbi)
 		//WRITE_VREG(HEVC_SAO_CTRL5, data32);
 		READ_WRITE_DATA16(pbi, HEVC_SAO_CTRL5, 0, 16, 8);
 	} else {
-		uint32_t data;
+		uint32_t data = 0;
 		if (get_cpu_major_id() >= AM_MESON_CPU_MAJOR_ID_T7)
 			WRITE_BACK_8(pbi, HEVC_SAO_CTRL26, 0);
 
diff --git a/common_drivers/drivers/media_modules/frame_provider/decoder_v4l/vav1_fb/av1_fb_hw.c b/common_drivers/drivers/media_modules/frame_provider/decoder_v4l/vav1_fb/av1_fb_hw.c
index d19029d3d..eb6d5cd03 100644
--- a/common_drivers/drivers/media_modules/frame_provider/decoder_v4l/vav1_fb/av1_fb_hw.c
+++ b/common_drivers/drivers/media_modules/frame_provider/decoder_v4l/vav1_fb/av1_fb_hw.c
@@ -2597,7 +2597,7 @@ static void config_sao_hw_fb(struct AV1HW_s *hw, param_t* params)
 		//WRITE_VREG(HEVC_SAO_CTRL5, data32);
 		READ_WRITE_DATA16(hw, HEVC_SAO_CTRL5, 0, 16, 8);
 	} else {
-		uint32_t data;
+		uint32_t data = 0;
 		if (get_cpu_major_id() >= AM_MESON_CPU_MAJOR_ID_T7) {
 			WRITE_BACK_8(hw, HEVC_SAO_CTRL26, 0);
 		}
diff --git a/common_drivers/drivers/media_modules/frame_provider/decoder_v4l/vp9_fb/vvp9_fb_v4l.c b/common_drivers/drivers/media_modules/frame_provider/decoder_v4l/vp9_fb/vvp9_fb_v4l.c
index ba3d9aa81..de00010d3 100644
--- a/common_drivers/drivers/media_modules/frame_provider/decoder_v4l/vp9_fb/vvp9_fb_v4l.c
+++ b/common_drivers/drivers/media_modules/frame_provider/decoder_v4l/vp9_fb/vvp9_fb_v4l.c
@@ -4581,7 +4581,7 @@ static void config_sao_hw_fb(struct VP9Decoder_s *pbi)
 		//WRITE_VREG(HEVC_SAO_CTRL5, data32);
 		READ_WRITE_DATA16(pbi, HEVC_SAO_CTRL5, 0, 16, 8);
 	} else {
-		uint32_t data;
+		uint32_t data = 0;
 		if (get_cpu_major_id() >= AM_MESON_CPU_MAJOR_ID_T7)
 			WRITE_BACK_8(pbi, HEVC_SAO_CTRL26, 0);
 
diff --git a/common_drivers/sound/soc/codecs/amlogic/Makefile b/common_drivers/sound/soc/codecs/amlogic/Makefile
index 128217abd..c511c0f30 100644
--- a/common_drivers/sound/soc/codecs/amlogic/Makefile
+++ b/common_drivers/sound/soc/codecs/amlogic/Makefile
@@ -2,7 +2,9 @@
 
 KBUILD_CFLAGS += -Wlarger-than=32768
 #KBUILD_CFLAGS += -Wstack-usage=1856
+ifndef CONFIG_CC_IS_CLANG
 KBUILD_CFLAGS += -Wno-error=stringop-overread
+endif
 
 # Amlogic
 MODULE_NAME_DUMMY = amlogic-snd-codec-dummy
@@ -89,4 +91,4 @@ PR_FMT_SY6026L = $(subst amlogic-,,$(MODULE_NAME_SY6026L))
 PR_FMT_DEFINE_SY6026L = "-Dpr_fmt(fmt)= \"[$(PR_FMT_SY6026L)]: \" fmt"
 obj-$(CONFIG_AMLOGIC_SND_SOC_SY6026L) += $(MODULE_NAME_SY6026L).o
 CFLAGS_SY6026L.o += $(PR_FMT_DEFINE_SY6026L)
-$(MODULE_NAME_SY6026L)-y += sy602x.o
\ No newline at end of file
+$(MODULE_NAME_SY6026L)-y += sy602x.o
diff --git a/crypto/aml_zstd.c b/crypto/aml_zstd.c
index 77977d9f5..59333c48d 100644
--- a/crypto/aml_zstd.c
+++ b/crypto/aml_zstd.c
@@ -11,7 +11,11 @@
 #include <linux/module.h>
 #include <linux/net.h>
 #include <linux/vmalloc.h>
+#ifdef CONFIG_AMLOGIC_ZSTD
+#include <linux/aml_zstd.h>
+#else
 #include <linux/zstd.h>
+#endif
 #include <crypto/internal/scompress.h>
 
 #ifdef CONFIG_AMLOGIC_ZSTD
diff --git a/drivers/gpu/arm/midgard/platform/devicetree/scaling.c b/drivers/gpu/arm/midgard/platform/devicetree/scaling.c
index be351ee9f..0e2a0fe74 100644
--- a/drivers/gpu/arm/midgard/platform/devicetree/scaling.c
+++ b/drivers/gpu/arm/midgard/platform/devicetree/scaling.c
@@ -147,7 +147,7 @@ static int gp_pll_user_cb_gpu(struct gp_pll_user_handle_s *user,
 
 int mali_perf_set_num_pp_cores(int cores)
 {
-    cores = cores;
+    //cores = cores;
     return 0;
 }
 
diff --git a/drivers/usb/gadget/function/f_hid.c b/drivers/usb/gadget/function/f_hid.c
index 6be6009f9..d7a949c03 100644
--- a/drivers/usb/gadget/function/f_hid.c
+++ b/drivers/usb/gadget/function/f_hid.c
@@ -79,6 +79,10 @@ struct f_hidg {
 	struct usb_ep			*out_ep;
 };
 
+#ifdef CONFIG_AMLOGIC_BRIDGE_HID
+struct f_hidg *hid_h;
+#endif
+
 static inline struct f_hidg *func_to_hidg(struct usb_function *f)
 {
 	return container_of(f, struct f_hidg, func);
@@ -519,6 +523,97 @@ static ssize_t f_hidg_write(struct file *file, const char __user *buffer,
 	return status;
 }
 
+#ifdef CONFIG_AMLOGIC_BRIDGE_HID
+ssize_t f_hidg_internal_write(char *buffer, size_t count, int nonblock)
+{
+	struct f_hidg *hidg  = hid_h;
+	struct usb_request *req;
+	unsigned long flags;
+	ssize_t status = -ENOMEM;
+
+	if (!hidg)
+		return status;
+
+	spin_lock_irqsave(&hidg->write_spinlock, flags);
+
+	if (!hidg->req) {
+		spin_unlock_irqrestore(&hidg->write_spinlock, flags);
+		return -ESHUTDOWN;
+	}
+
+#define WRITE_COND (!hidg->write_pending)
+try_again:
+	/* write queue */
+	while (!WRITE_COND) {
+		spin_unlock_irqrestore(&hidg->write_spinlock, flags);
+		if (nonblock & O_NONBLOCK)
+			return -EAGAIN;
+
+		if (wait_event_interruptible_exclusive(hidg->write_queue, WRITE_COND))
+			return -ERESTARTSYS;
+
+		spin_lock_irqsave(&hidg->write_spinlock, flags);
+	}
+
+	hidg->write_pending = 1;
+	req = hidg->req;
+	count  = min_t(unsigned int, count, hidg->report_length);
+
+	spin_unlock_irqrestore(&hidg->write_spinlock, flags);
+
+	if (!req) {
+		ERROR(hidg->func.config->cdev, "hidg->req is NULL\n");
+		status = -ESHUTDOWN;
+		goto release_write_pending;
+	}
+
+	memcpy(req->buf, buffer, count);
+
+	spin_lock_irqsave(&hidg->write_spinlock, flags);
+
+	/* when our function has been disabled by host */
+	if (!hidg->req) {
+		free_ep_req(hidg->in_ep, req);
+		/*
+		 * TODO
+		 * Should we fail with error here?
+		 */
+		goto try_again;
+	}
+
+	req->status   = 0;
+	req->zero     = 0;
+	req->length   = count;
+	req->complete = f_hidg_req_complete;
+	req->context  = hidg;
+
+	spin_unlock_irqrestore(&hidg->write_spinlock, flags);
+
+	if (!hidg->in_ep->enabled) {
+		ERROR(hidg->func.config->cdev, "in_ep is disabled\n");
+		status = -ESHUTDOWN;
+		goto release_write_pending;
+	}
+
+	status = usb_ep_queue(hidg->in_ep, req, GFP_ATOMIC);
+	if (status < 0)
+		goto release_write_pending;
+	else
+		status = count;
+
+	return status;
+release_write_pending:
+	spin_lock_irqsave(&hidg->write_spinlock, flags);
+	hidg->write_pending = 0;
+	spin_unlock_irqrestore(&hidg->write_spinlock, flags);
+
+	wake_up(&hidg->write_queue);
+
+	return status;
+}
+EXPORT_SYMBOL(f_hidg_internal_write);
+#endif
+
 static __poll_t f_hidg_poll(struct file *file, poll_table *wait)
 {
 	struct f_hidg	*hidg  = file->private_data;
@@ -1009,6 +1104,10 @@ static int hidg_bind(struct usb_configuration *c, struct usb_function *f)
 	if (status)
 		goto fail_free_descs;
 
+#ifdef CONFIG_AMLOGIC_BRIDGE_HID
+	hid_h = hidg;
+#endif
+
 	return 0;
 fail_free_descs:
 	usb_free_all_descriptors(f);
@@ -1250,6 +1349,10 @@ static void hidg_unbind(struct usb_configuration *c, struct usb_function *f)
 {
 	struct f_hidg *hidg = func_to_hidg(f);
 
+#ifdef CONFIG_AMLOGIC_BRIDGE_HID
+	hid_h = NULL;
+#endif
+
 	cdev_device_del(&hidg->cdev, &hidg->dev);
 
 	usb_free_all_descriptors(f);
diff --git a/fs/btrfs/aml_zstd.c b/fs/btrfs/aml_zstd.c
index fc42dd0ba..6ce266847 100644
--- a/fs/btrfs/aml_zstd.c
+++ b/fs/btrfs/aml_zstd.c
@@ -16,7 +16,11 @@
 #include <linux/refcount.h>
 #include <linux/sched.h>
 #include <linux/slab.h>
+#ifdef CONFIG_AMLOGIC_ZSTD
+#include <linux/aml_zstd.h>
+#else
 #include <linux/zstd.h>
+#endif
 #include "misc.h"
 #include "compression.h"
 #include "ctree.h"
diff --git a/fs/f2fs/aml_super.c b/fs/f2fs/aml_super.c
index 100a27121..519dca214 100644
--- a/fs/f2fs/aml_super.c
+++ b/fs/f2fs/aml_super.c
@@ -25,7 +25,11 @@
 #include <linux/quota.h>
 #include <linux/unicode.h>
 #include <linux/part_stat.h>
+#ifdef CONFIG_AMLOGIC_ZSTD
+#include <linux/aml_zstd.h>
+#else
 #include <linux/zstd.h>
+#endif
 #include <linux/lz4.h>
 
 #include "f2fs.h"
diff --git a/fs/incfs/aml_data_mgmt.h b/fs/incfs/aml_data_mgmt.h
index 2ac496410..3960abfd9 100644
--- a/fs/incfs/aml_data_mgmt.h
+++ b/fs/incfs/aml_data_mgmt.h
@@ -14,7 +14,11 @@
 #include <linux/rcupdate.h>
 #include <linux/completion.h>
 #include <linux/wait.h>
+#ifdef CONFIG_AMLOGIC_ZSTD
+#include <linux/aml_zstd.h>
+#else
 #include <linux/zstd.h>
+#endif
 #include <crypto/hash.h>
 #include <linux/rwsem.h>
 
diff --git a/fs/squashfs/aml_zstd_wrapper.c b/fs/squashfs/aml_zstd_wrapper.c
index c40445dbf..4126cceb7 100644
--- a/fs/squashfs/aml_zstd_wrapper.c
+++ b/fs/squashfs/aml_zstd_wrapper.c
@@ -11,7 +11,11 @@
 #include <linux/mutex.h>
 #include <linux/bio.h>
 #include <linux/slab.h>
+#ifdef CONFIG_AMLOGIC_ZSTD
+#include <linux/aml_zstd.h>
+#else
 #include <linux/zstd.h>
+#endif
 #include <linux/vmalloc.h>
 
 #include "squashfs_fs.h"
diff --git a/include/linux/usb/g_hid.h b/include/linux/usb/g_hid.h
index 7581e488c..5615c335d 100644
--- a/include/linux/usb/g_hid.h
+++ b/include/linux/usb/g_hid.h
@@ -30,4 +30,7 @@ struct hidg_func_descriptor {
 	unsigned char		report_desc[];
 };
 
+#ifdef CONFIG_AMLOGIC_BRIDGE_HID
+ssize_t f_hidg_internal_write(char *buffer, size_t count, int nonblock);
+#endif
 #endif /* __LINUX_USB_G_HID_H */
